pkg/authorization/api/conversion.go

```diff
@@ -0,0 +1,202 @@
+package api
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/user/api/validation"
+)
+
+func addConversionFuncs(scheme *runtime.Scheme) error {
+	if err := scheme.AddConversionFuncs(
+		Convert_api_ClusterRole_To_rbac_ClusterRole,
+		Convert_api_Role_To_rbac_Role,
+		Convert_api_ClusterRoleBinding_To_rbac_ClusterRoleBinding,
+		Convert_api_RoleBinding_To_rbac_RoleBinding,
+		Convert_rbac_ClusterRole_To_api_ClusterRole,
+		Convert_rbac_Role_To_api_Role,
+		Convert_rbac_ClusterRoleBinding_To_api_ClusterRoleBinding,
+		Convert_rbac_RoleBinding_To_api_RoleBinding,
+	); err != nil { // If one of the conversion functions is malformed, detect it immediately.
+		return err
+	}
+	return nil
+}
+
+func Convert_api_ClusterRole_To_rbac_ClusterRole(in *ClusterRole, out *rbac.ClusterRole, _ conversion.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+	out.Rules = convert_api_PolicyRules_To_rbac_PolicyRules(in.Rules)
+	return nil
+}
+
+func Convert_api_Role_To_rbac_Role(in *Role, out *rbac.Role, _ conversion.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+	out.Rules = convert_api_PolicyRules_To_rbac_PolicyRules(in.Rules)
+	return nil
+}
+
+func Convert_api_ClusterRoleBinding_To_rbac_ClusterRoleBinding(in *ClusterRoleBinding, out *rbac.ClusterRoleBinding, _ conversion.Scope) error {
+	var err error
+	if out.Subjects, err = convert_api_Subjects_To_rbac_Subjects(in.Subjects); err != nil {
+		return err
+	}
+	out.RoleRef = convert_api_RoleRef_To_rbac_RoleRef(&in.RoleRef)
+	out.ObjectMeta = in.ObjectMeta
+	return nil
+}
+
+func Convert_api_RoleBinding_To_rbac_RoleBinding(in *RoleBinding, out *rbac.RoleBinding, _ conversion.Scope) error {
+	if len(in.RoleRef.Namespace) != 0 && in.RoleRef.Namespace != in.Namespace {
+		return fmt.Errorf("invalid origin role binding %s: attempts to reference role in namespace %q instead of current namespace %q", in.Name, in.RoleRef.Namespace, in.Namespace)
+	}
+	var err error
+	if out.Subjects, err = convert_api_Subjects_To_rbac_Subjects(in.Subjects); err != nil {
+		return err
+	}
+	out.RoleRef = convert_api_RoleRef_To_rbac_RoleRef(&in.RoleRef)
+	out.ObjectMeta = in.ObjectMeta
+	return nil
+}
+
+func convert_api_PolicyRules_To_rbac_PolicyRules(in []PolicyRule) []rbac.PolicyRule {
+	rules := make([]rbac.PolicyRule, 0, len(in))
+	for _, rule := range in {
+		r := rbac.PolicyRule{ // AttributeRestrictions is lost, but our authorizor ignores that field now
```

Comment:
if we see a rule with that restriction, we need to limit the rule to the things that restriction would allow, and convert subjectaccessreviews resources to selfsubjectaccessreviews.authorization.k8s.io resources